topic-operator/src/main/java/io/strimzi/operator/topic/v2/BatchingLoop.java

```diff
@@ -0,0 +1,302 @@
+                LOGGER.debugOp("[Batch #{}] contains {}", batchId, r);
+                return new ReconcilableTopic(r, kt, BatchingTopicController.topicName(kt));
+            } else {
+                // Null can happen if the KafkaTopic has been deleted from Kube and we've not yet processed
+                // the corresponding delete event
+                LOGGER.traceOp("[Batch #{}] Lookup from item store for {} yielded nothing",
+                        batchId, topicUpsert);
+                return null;
+            }
+        }
+
+        private void fillBatch(int batchId, Batch batch) throws InterruptedException {
+            LOGGER.infoOp("[Batch #{}] Filling", batchId);
+            List<TopicEvent> rejected = new ArrayList<>();
+
+
+            final long deadlineNanoTime = System.nanoTime() + maxBatchLingerMs * 1_000_000;
+            while (true) {
+                if (batch.size() >= maxBatchSize) {
+                    LOGGER.traceOp("[Batch #{}] reached maxBatchSize, batch complete", batchId, maxBatchSize);
+                    break;
+                }
+
+                long timeoutNs = deadlineNanoTime - System.nanoTime();
+                if (timeoutNs <= 0) {
+                    LOGGER.traceOp("[Batch #{}] {}ms linger expired", batchId, maxBatchLingerMs);
+                    break;
+                }
+                LOGGER.traceOp("[Batch #{}] Taking next item from deque head with timeout {}ns", batchId, timeoutNs);
+                TopicEvent topicEvent = queue.pollFirst(timeoutNs, TimeUnit.NANOSECONDS);
+
+                if (topicEvent == null) {
+                    LOGGER.traceOp("[Batch #{}] linger expired, batch complete", batchId);
+                    break;
+                }
+                addToBatch(batchId, batch, rejected, topicEvent);
+            }
+            LOGGER.traceOp("[Batch #{}] Filled with {} topics", batchId, batch.size());
+
+            // here we need a deque and can push `rejected` back on the front of the queue
+            //      where they can be taken by the next thread.
+            for (int i = rejected.size() - 1; i >= 0; i--) {
+                TopicEvent item = rejected.get(i);
+                if (queue.offerFirst(item)) {
+                    LOGGER.traceOp("[Batch #{}] Item {} returned to deque head", batchId, item);
+                } else {
+                    LOGGER.warnOp("[Batch #{}] Item {} couldn't be pushed onto the deque head after rejection from batch", batchId, item);
+                }
+            }
+        }
+
+        private void addToBatch(int batchId, Batch batch, List<TopicEvent> rejected, TopicEvent topicEvent) {
+            // We could add logic here to cope properly with interleaved upserts and deletes
+            // of the same topic in the same batch.
+            // E.g. upset then delete is equivalent to just a delete
+            // It's actually a bit tricky since you have to process the events in reverse order to correctly
```

Comment:
There are pros and cons. I've thought about it far too hard. 

Pros:

* It's dead simple
* It means we don't really need fiddly tests to cover the cases

Cons: 

* It means we have to use a deque, which might not perform great
* It can increase the latency for processing many events happening to the same topic

Note that because the deque/queue contains events, not simply topics, we can't easily put a logically correct upper bound on how big it should get (unlike the UserOperator, where the queue holds _unique_ `KafkaUsers`, so you only need a queue that's as big as the number of users you expect). So the latency issue could actually manifest as `offer` failing, and incorrect behaviour (e.g. missed deletion events when finalizers are not in use resulting in topics remaining in Kafka). It would probably be best to exit the whole TO process in that case, forcing the user to increase the size of the deque. Wdyt?

Reply:
> It would probably be best to exit the whole TO process in that case, forcing the user to increase the size of the deque. Wdyt?

I agree with that. The termination message should be something like "Insufficient event queue capacity, please raise by setting STRIMZI_MAX_QUEUE_SIZE environment variable".

That said, the STRIMZI_MAX_QUEUE_SIZE default value seems a bit low thinking about a busy cluster. Should be align with what we have on the user operator (1024)? We should also enforce STRIMZI_MAX_QUEUE_SIZE >= STRIMZI_MAX_BATCH_SIZE.

Reply:
Bumped up the default size to 1024 as you suggest. `STRIMZI_MAX_QUEUE_SIZE >= STRIMZI_MAX_BATCH_SIZE` doesn't necessarily make sense, since the adding to the quieue and the batch creation happen in different threads, so you could actually have a bigger batch than queue and it would work, so long as the queue offer was not rejected.

I quite like the idea of using a termination message (written in `/dev/termination-log`, per [kube defaults](https://kubernetes.io/docs/tasks/debug/debug-application/determine-reason-pod-failure/)). But we'd need a mechanism to be able to test it without actually exiting the test JVM. Also interested in what @scholzj and @ppatierno think about this.

Reply:
I think simply logging it will be equally if not more effective and a lot less trouble ... unless you plan any automated processing of the error message.

Reply:
Do you mean logging prior to exit (as opposed to a proper termination message)? Or logging and carrying on?

Reply:
I meant mainly the use of the termination log. TBH, I do not understand the issue well enough to comment on whether the operator should exit or not. I think that depends on:
* Whether it can recover
* How much you want the users to notice it (they are obviously much more likely to notice it when it starts crash looping -> but might not really care if it happens only once per week due to some run of events.). But once it would be restarting too often, the regular log is IMHO fine to find the cause.

I think the termination message might be useful especially if for example we have some further tooling to monitor the TO and suggest some insights to reconfigure it.